lms/djangoapps/verify_student/views.py

```diff
@@ -269,36 +269,45 @@ def get(
+        all_modes, course_mode, expired_course_mode = self._get_all_and_verified_modes_for_course(course_key)
 
-        # Check that the course has an unexpired verified mode
-        course_mode, expired_course_mode = self._get_verified_modes_for_course(course_key)
-
-        if course_mode is not None:
-            log.info(
-                u"Entering verified workflow for user '%s', course '%s', with current step '%s'.",
-                request.user.id, course_id, current_step
-            )
-        elif expired_course_mode is not None:
-            # Check if there is an *expired* verified course mode;
-            # if so, we should show a message explaining that the verification
-            # deadline has passed.
-            log.info(u"Verification deadline for '%s' has passed.", course_id)
-            context = {
-                'course': course,
-                'deadline': (
-                    get_default_time_display(expired_course_mode.expiration_datetime)
-                    if expired_course_mode.expiration_datetime else ""
+        # Check that the course has an unexpired no-id-professional mode
+        no_id_prof_mode = CourseMode.mode_for_course(
+            course_key,
+            CourseMode.NO_ID_PROFESSIONAL_MODES[0],
+            modes=all_modes
+        )
+        # If there is no no-id-professional mode then check the verified track modes
+        if not no_id_prof_mode:
+            # Check that the course has an unexpired verified mode
+            if course_mode is not None:
+                log.info(
+                    u"Entering verified workflow for user '%s', course '%s', with current step '%s'.",
+                    request.user.id, course_id, current_step
                 )
-            }
-            return render_to_response("verify_student/missed_verification_deadline.html", context)
+            elif expired_course_mode is not None:
+                # Check if there is an *expired* verified course mode;
+                # if so, we should show a message explaining that the verification
+                # deadline has passed.
+                log.info(u"Verification deadline for '%s' has passed.", course_id)
+                context = {
+                    'course': course,
+                    'deadline': (
+                        get_default_time_display(expired_course_mode.expiration_datetime)
+                        if expired_course_mode.expiration_datetime else ""
+                    )
+                }
+                return render_to_response("verify_student/missed_verification_deadline.html", context)
+            else:
+                # Otherwise, there has never been a verified mode,
+                # so return a page not found response.
+                log.warn(
+                    u"No verified course mode found for course '%s' for verification flow request",
+                    course_id
+                )
+                raise Http404
```

Comment:
I'm still concerned that this code is making too many assumptions about which course modes are defined and what their properties are.

In the short term, we can make the assumption that there is exactly on paid mode for a course, which can either require verification or not.  If we make this assumption, the logic here becomes much simpler:

1) Retrieve the paid course mode (using `paid_modes_for_course()`).  If there is more than one, log an error (since this is likely a misconfiguration), then choose the first one in the list.

2) If no paid course mode is found, then check if there is an expired verified mode, and show the "missed verification deadline" message.

3) If the paid course mode is verified, include the verification steps.  Otherwise, do NOT include the verification steps.